iceoryx_posh/source/roudi/service_registry.cpp

```diff
@@ -17,45 +17,149 @@
+    m_instanceMap.insert({serviceDescription.getInstanceIDString(), m_serviceDescriptionVector.size() - 1});
+    return cxx::success<>();
 }
 
-void ServiceRegistry::remove(const capro::IdString_t& service, const capro::IdString_t& instance)
+bool ServiceRegistry::remove(const capro::ServiceDescription& serviceDescription) noexcept
 {
-    cxx::set::remove(m_serviceMap[service].instanceSet, instance);
+    if (!serviceDescription.isValid())
+    {
+        return false;
+    }
+
+    bool removedElement{false};
+    uint64_t index = 0U;
+    for (auto iterator = m_serviceDescriptionVector.begin(); iterator != m_serviceDescriptionVector.end();)
+    {
+        if (m_serviceDescriptionVector[index] == serviceDescription)
+        {
+            m_serviceDescriptionVector.erase(iterator);
+            removedElement = true;
+            // There can be not more than one element
+            break;
+        }
+        index++;
+        iterator++;
+    }
+
+    auto removeIndexFromMap = [](std::multimap<capro::IdString_t, uint64_t>& map, uint64_t index) {
+        for (auto it = map.begin(); it != map.end();)
+        {
+            if (it->second == index)
+            {
+                it = map.erase(it);
+                continue;
+            }
+            else if (it->second > index)
+            {
+                // update index due to removed element
+                it->second--;
+            }
+            it++;
+        }
+    };
+
+    removeIndexFromMap(m_serviceMap, index);
+    removeIndexFromMap(m_instanceMap, index);
+
+    return removedElement;
 }
 
-void ServiceRegistry::find(InstanceSet_t& instances,
+void ServiceRegistry::find(ServiceDescriptionVector_t& searchResult,
                            const capro::IdString_t& service,
-                           const capro::IdString_t& instance) const
+                           const capro::IdString_t& instance) const noexcept
 {
-    if (instance == Wildcard)
+    cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> intersection;
+
+    // Find (K1, K2)
+    // O(log n + log n + max(#PossibleServices + #possiblesInstances) + #intersection)
+    if (instance != Wildcard && service != Wildcard)
+    {
+        cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> possibleServices;
+        cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> possibleInstances;
+
+        auto rangeServiceMap = m_serviceMap.equal_range(service);
+        for (auto entry = rangeServiceMap.first; entry != rangeServiceMap.second; ++entry)
+        {
+            possibleServices.push_back(entry->second);
+        }
+
+        auto rangeInstanceMap = m_instanceMap.equal_range(instance);
+        for (auto entry = rangeInstanceMap.first; entry != rangeInstanceMap.second; ++entry)
+        {
+            possibleInstances.push_back(entry->second);
+        }
+
+        ::std::set_intersection(possibleServices.begin(),
```

Comment:
Note that they need to be sorted before. They should be, with the ranges coming from an ordered map.